From cd180afba3cbc466ec906ed9bb4a8f6a9ed84d50 Mon Sep 17 00:00:00 2001
From: Zheng Yejian <zhengyejian1@huawei.com>
Date: Wed, 7 Sep 2022 17:20:22 +0000
Subject: [PATCH 0204/1076] [Huawei] livepatch: Complete check calltrace for
 running tasks

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187517

--------------------------------

As before, task calltrace checking was assumed being performed under
stop_machine and tasks to be checked are all not in running state.
So there is one case that old function not needed to check, that is:
  preemption disabled and 'force' field not be KLP_NORMAL_FORCE and
  no 'call' instructions in livepatch replace area.

But when using breakpoint optimization without stop_machine, tasks
may be running, we can not ignore above check.

Signed-off-by: Zheng Yejian <zhengyejian1@huawei.com>
---
 arch/arm/kernel/livepatch.c        | 1 +
 arch/arm64/kernel/livepatch.c      | 1 +
 arch/powerpc/kernel/livepatch_32.c | 1 +
 arch/powerpc/kernel/livepatch_64.c | 1 +
 arch/x86/kernel/livepatch.c        | 1 +
 5 files changed, 5 insertions(+)

diff --git a/arch/arm/kernel/livepatch.c b/arch/arm/kernel/livepatch.c
index d5d94593012a..9afef2f32edc 100644
--- a/arch/arm/kernel/livepatch.c
+++ b/arch/arm/kernel/livepatch.c
@@ -133,6 +133,7 @@ int arch_klp_check_activeness_func(struct klp_patch *patch, int enable,
 				 * complete.
 				 */
 				if (IS_ENABLED(CONFIG_PREEMPTION) ||
+				    IS_ENABLED(CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE) ||
 				    (func->force == KLP_NORMAL_FORCE) ||
 				    check_jump_insn(func_addr)) {
 					ret = add_func(func_list, &pcheck,
diff --git a/arch/arm64/kernel/livepatch.c b/arch/arm64/kernel/livepatch.c
index 3805e4054d4f..e23e6597fc1e 100644
--- a/arch/arm64/kernel/livepatch.c
+++ b/arch/arm64/kernel/livepatch.c
@@ -122,6 +122,7 @@ int arch_klp_check_activeness_func(struct klp_patch *patch, int enable,
 				 * complete.
 				 */
 				if (IS_ENABLED(CONFIG_PREEMPTION) ||
+				    IS_ENABLED(CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE) ||
 				    (func->force == KLP_NORMAL_FORCE) ||
 				    check_jump_insn(func_addr)) {
 					ret = add_func(func_list, &pcheck,
diff --git a/arch/powerpc/kernel/livepatch_32.c b/arch/powerpc/kernel/livepatch_32.c
index 0018e9c420af..b6bba306d525 100644
--- a/arch/powerpc/kernel/livepatch_32.c
+++ b/arch/powerpc/kernel/livepatch_32.c
@@ -123,6 +123,7 @@ int arch_klp_check_activeness_func(struct klp_patch *patch, int enable,
 				 * complete.
 				 */
 				if (IS_ENABLED(CONFIG_PREEMPTION) ||
+				    IS_ENABLED(CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE) ||
 				    (func->force == KLP_NORMAL_FORCE) ||
 				    check_jump_insn(func_addr)) {
 					ret = add_func(func_list, &pcheck,
diff --git a/arch/powerpc/kernel/livepatch_64.c b/arch/powerpc/kernel/livepatch_64.c
index c4bce597b69b..2e9a89a0edfc 100644
--- a/arch/powerpc/kernel/livepatch_64.c
+++ b/arch/powerpc/kernel/livepatch_64.c
@@ -131,6 +131,7 @@ int arch_klp_check_activeness_func(struct klp_patch *patch, int enable,
 				 * complete.
 				 */
 				if (IS_ENABLED(CONFIG_PREEMPTION) ||
+				    IS_ENABLED(CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE) ||
 				    (func->force == KLP_NORMAL_FORCE) ||
 				    check_jump_insn(func_addr)) {
 					ret = add_func(func_list, &pcheck,
diff --git a/arch/x86/kernel/livepatch.c b/arch/x86/kernel/livepatch.c
index 6c5ad1a82ab0..4c8c1f0802ea 100644
--- a/arch/x86/kernel/livepatch.c
+++ b/arch/x86/kernel/livepatch.c
@@ -120,6 +120,7 @@ int arch_klp_check_activeness_func(struct klp_patch *patch, int enable,
 				 * complete.
 				 */
 				if (IS_ENABLED(CONFIG_PREEMPTION) ||
+				    IS_ENABLED(CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE) ||
 				    (func->force == KLP_NORMAL_FORCE) ||
 				    check_jump_insn(func_addr)) {
 					ret = add_func(func_list, &pcheck,
-- 
2.17.1

